chemprop/models/ffn.py

```diff
@@ -0,0 +1,350 @@
+                        dropout=dropout,
+                        activation=activation,
+                    ),
+                )
+            else:
+                self.ffn_readout = build_ffn(
+                    first_linear_dim=features_size,
+                    hidden_size=hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                )
+        self.constraint = constraint
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.Tensor, List, torch.Tensor, List, torch.Tensor],
+        constraints: torch.Tensor,
+        bond_types: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise ValueError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if constraints is not None:
+            if self.ffn_type == "atom":
+                output_hidden = self.ffn(hidden)
+                output = self.ffn_readout(output_hidden)
+            elif self.ffn_type == "bond":
+                b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+                b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+                output_1 = self.ffn(b_hidden_1)
+                output_2 = self.ffn(b_hidden_2)
+                output_hidden = (output_1 + output_2) / 2
+                output = self.ffn_readout(output_hidden)
+                if bond_types is not None:
+                    output = output + bond_types.reshape(-1, 1)
+
+            u_a = self.weights_readout(output_hidden)
+            constrained_output = []
+            for i, (start, size) in enumerate(scope):
+                if size == 0:
+                    continue
+                else:
+                    u_ai = u_a[start:start+size]
+                    qi = output[start:start+size]
+                    Q = constraints[i]
+
+                    wi = F.softmax(u_ai, dim=0).flatten()
+
+                    qi[:, 0] = qi[:, 0] + wi * (Q - qi[:, 0].sum())
+                    constrained_output.append(qi)
```

Comment:
do we need to do all this indexing and slicing and flattening? seems messy and i think broadcasting will work just fine here

Reply:
I think the slicing and indexing are necessary because the softmax will act on the $u\hat{a_i}$  of atoms in a molecule. The purpose of flattening the output of softmax is to let the shape of `wi` be the same as  `qi[:, 0]`. What do you think about using broadcasting here?

Reply:
`wi`: (number_of_atoms)
`qi`: (number_of_atoms, 1)
`u_ai `: (number_of_atoms, 1)
`output`: (sum_of_number_of_atoms, 1)

PS. The `sum_of_number_of_atoms` is the sum of number of atoms for each molecule in this batch.

Reply:
```python
>>> n_atoms = 10
>>> Q = 1
>>> q_i = torch.rand(n_atoms, 1)
>>> u_ai = torch.rand(n_atoms, 1)
>>> w_i = u_ai.softmax(0)
# your code
>>> # separating the assignment into two steps
>>> q_f1 = q_i[:, 0] + w_i.flatten() * (Q - q_i[:, 0].sum())
>>> q_f1.shape
torch.Size([10])
>>> # assignment would happen here: q_i[:, 0] = q_f1
# suggestion
>>> q_f = q_i + w_i * (Q - q_i.sum())
>>> q_f.shape
torch.Size([10, 1])
>>> q_f2[:, 0].equal(q_f1)
True
>>> constrained_output.append(q_f2)